build: apk-related improvements

GeorgeSapkin opened the pull request:
Allow defining virtual provides using the `PROVIDES` field by prefixing them with an `@`, e.g.:

```
PROVIDES:=@ca-certs
```

Virtual provides don't own the provided name and multiple packages with the same virtual provides can be installed side-by-side. Packages must still take care not to override each other's files.

Add an implicit self-provide to packages. apk can't handle self provides, be it versioned or virtual, so opt for a suffix instead. This allows several variants to provide the same virtual package without adding extra provides to the default one, e.g. `wget` implicitly provides `wget-any` and is marked as default, so `wget-ssl` can explicitly provide `@wget-any` as well.

Add an implicit self-provide to kmods using a prefix and a suffix. Package name without a prefix/suffix is too generic and might conflict with other packages, e.g. `wireguard`. This allows several variants to provide the same virtual package without adding extra provides to the default one, e.g. `r8169` implicitly provides `kmod-r8169-any` and is marked as default, so `r8125` can explicitly provide `@kmod-r8169-any` as well.

The actual symbol that marks virtual provides is not important. I chose @ for _alias_.

Edit: @ is kind of overloaded as it already means something else in `DEPENDS`.

> [!IMPORTANT]
>
> Cross-package provides must be virtual and a default variant must be set.

Refactor provides logic into a helper define and use it for both apk and control (which was still using the older logic). Document the behavior. It looks excessive but I think we should have this somewhere outside of commit messages.

Filter out virtual provides when generating metadata.

Filter out virtual provides prefix and self provide where appropriate.

Remove unnecessary logging.

Adjust packages to uses the new virtual provides:

- Switch ca-certs provides to use the new virtual provides semantic that enables ca-bundle and ca-certificates to be installed side-by-side. Provide the new format virtual `ca-certificates-any` in ca-bundle.
- Switch ath10k and related kmods to use the new virtual kmod provides semantic and mark ath10k as the default variant.
- Switch r8169 and related kmods to use the new virtual kmod provides semantic and mark r8169 as the default variant.
- Switch rtl8812au-ct to use the new virtual kmod provides semantic.

Tested:
- building and installing ca-bundle and ca-certificates side-by-side
- building and installing stubby that depends on the virtual ca-certs
- building and installing libsqlite3 that has an ABI and sqlite3-cli that depends on it
- building x86/64 generic image and testing in QEMU
- defining a provides in a library with an ABI
- build config can still be generated with ca-bundle and ca-certificates selected at the same time (this might need more testing)

The alternatives check might not be necessary anymore and could be replaced with the new virtual semantic in the long run, but is there for compatibility reasons.

> [!CAUTION]
>
> This was neither tested with, nor is intended to be backported to opkg.

Fixes: https://github.com/openwrt/openwrt/issues/21257

cc: @efahl, @robimarko and maybe @mstorchak 

'Tis the holiday season over here, so I probably overlooked something obvious. I think this needs a good look before this gets merged. I will update the wiki to document the recent DEPENDS, EXTRA_DEPENDS and PROVIDES changes at some point, unless somebody else gets to it first.

GeorgeSapkin committed:
build: refactor provides logic

Refactor provides logic into a helper define and use it for both apk and
control. Document the behavior.

Store preformatted provides in Package/$(1)/PROVIDES similar to defines.

Remove unnecessary logging.

Signed-off-by: George Sapkin <george@sapk.in>
Link: https://github.com/openwrt/openwrt/pull/21288
Signed-off-by: Robert Marko <robimarko@gmail.com>

GeorgeSapkin committed:
build: add support for virtual provides

Allow defining virtual provides using the PROVIDES field by prefixing
them with @, e.g.:

PROVIDES:=@ca-certs

Virtual provides don't own the provided name and multiple packages with
the same virtual provides can be installed side-by-side. Packages must
still take care not to override each other's files.

Add an implicit self-provide to packages. apk can't handle self
provides, be it versioned or virtual, so opt for a suffix instead. This
allows several variants to provide the same virtual package without
adding extra provides to the default one, e.g. wget implicitly provides
wget-any and is marked as default, so wget-ssl can explicitly provide
@wget-any as well.

Filter out virtual provides when generating metadata.

Filter out virtual provides prefix and self provide where appropriate.

Signed-off-by: George Sapkin <george@sapk.in>
Link: https://github.com/openwrt/openwrt/pull/21288
Signed-off-by: Robert Marko <robimarko@gmail.com>

GeorgeSapkin committed:
ca-certificates: provide a virtual package

Switch ca-certs provides to use the new virtual provides semantic that
enables ca-bundle and ca-certificates to be installed side-by-side.

Provide the new format virtual ca-certificates-any in ca-bundle.

Fixes: https://github.com/openwrt/openwrt/issues/21257
Signed-off-by: George Sapkin <george@sapk.in>
Link: https://github.com/openwrt/openwrt/pull/21288
Signed-off-by: Robert Marko <robimarko@gmail.com>

GeorgeSapkin commented:
@efahl I haven't seen this. I don't think it's relevant, because the @ part is being taken care of inside the build tools and doesn't reach apk. The actual depends that packages use doesn't have an @, so:

```Makefile
PROVIDES:=@ca-certs

...

DEPENDS:=+ca-certs
```

The @ is a _hint_, if you will, for the build tools to treat that provides differently.

map-b commented:
Have you thought about just using ALTERNATIVES with those two packages??  I'm also thinking if we need to reflect the ALTERNATIVES priority into --provides-priority somehow since the DEFAULT_VARIANT only covers the case when two providers are in one common package makefile source file?

GeorgeSapkin commented:
> Have you thought about just using ALTERNATIVES with those two packages?? I'm also thinking if we need to reflect the ALTERNATIVES priority into --provides-priority somehow since the DEFAULT_VARIANT only covers the case when two providers are in one common package makefile source file?

AFAICT, `ALTERNATIVES` are for executables, or at least standalone files, e.g. from wget:

```Makefile
ALTERNATIVES:=400:/usr/bin/wget:/usr/libexec/wget-ssl

...

ALTERNATIVES:=300:/usr/bin/wget:/usr/libexec/wget-nossl
```

ca-bundle and ca-certificates contain a distinct list of certificate files. Unless you mean something else.

efahl commented:
> The @ is a _hint_, if you will, for the build tools to treat that provides differently.

👍🏼 I'm considering their `abuild` as Alpine's rough equivalent to our `package-pack.mk`, and given that...

The way I read their script is that they're doing the same thing wrt metachars in the `provides`, so our use is very much in the same spirit, even if semantically disjoint.

GeorgeSapkin commented on include/package-pack.mk:
```diff
@@ -78,6 +78,62 @@ define FixupDependencies
+# Format provides both for apk and control
+#
+# - If ABI version is defined:
+#   - package is named `package_name-ABI_version`
```
This is also not entirely correct as it depends on whether a package ends in a number or not, e.g. `libsqlite3-0` where 0 is the ABI version. But I can certainly phrase it better.

efahl replied:
That one's is semi broken in its makefile:

```
$ apk query --fields tags libsqlite\*
Tags: openwrt:abiversion=-0   😂
```

efahl replied:
I believe that's an artifact of some ancient code that did ABI-sanitizing by naively stripping `[0-9.]` from the package name.  I think that's all fixed now, specifically `rpc-sys packagelist`, make-index-json.py and ipkg-remove all use the actual ABI tag to do the sanitizing.

efahl replied:
Well, if I did it, then it must be correct...  Right? 😅

But, in any case, it does makes the versioned names more readable, so I don't see a reason to change it back.

GeorgeSapkin replied:
It's creating an `index.json` that I _think_ is being used in asu. But what's the use case for specifying packages without an ABI version? Also if the tooling needs the package name without an ABI version, why not store that instead of doing the stripping magic? I mean the value is right there.

efahl replied:
Yes, the `make-index-json.py` script is primarily for ASU use, but some other tools make use of it, too.

`owut` (and previously `auc`) use the sanitized names in order to check if there's a newer package version (in order to decide if you need an upgrade or not).  There's no reliable way to correlate two versions if their names are ABI-versioned, so the rpc-sys code did the sanitizing historically, and the non-ABI names have become the convention across all of the ASU ecosystem.

GeorgeSapkin replied:
I clarified the ABI suffix part.

GeorgeSapkin commented:
@map-b I'm still not following what you want to put into alternatives. ca-certificates is not one file:

```bash
root@OpenWrt:~# apk query --fields=contents ca-bundle ca-certificates
Contents:
  etc/ssl/cert.pem
  etc/ssl/certs/ca-certificates.crt
  lib/apk/packages/ca-bundle.list

Contents:
  etc/ssl/certs/002c0b4f.0
  etc/ssl/certs/0179095f.0
  etc/ssl/certs/02265526.0
  etc/ssl/certs/062cdee6.0
  etc/ssl/certs/064e0aa9.0
  etc/ssl/certs/06dc52d5.0
  etc/ssl/certs/09789157.0
  etc/ssl/certs/0a775a30.0
  etc/ssl/certs/0b1b94ef.0
  etc/ssl/certs/0b9bc432.0
  etc/ssl/certs/0bf05006.0
  etc/ssl/certs/0f5dc4f3.0
  etc/ssl/certs/0f6fa695.0
  etc/ssl/certs/1001acf7.0
  etc/ssl/certs/106f3e4d.0
  etc/ssl/certs/14bc7599.0
  etc/ssl/certs/1cef98f5.0
  etc/ssl/certs/1d3472b9.0
  etc/ssl/certs/1e08bfd1.0
  etc/ssl/certs/1e09d511.0
  etc/ssl/certs/228f89db.0
  etc/ssl/certs/244b5494.0
  etc/ssl/certs/2923b3f9.0
  etc/ssl/certs/2ae6433e.0
  etc/ssl/certs/2b349938.0
  etc/ssl/certs/32888f65.0
  etc/ssl/certs/3513523f.0
  etc/ssl/certs/3bde41ac.0
  etc/ssl/certs/3e359ba6.0
  etc/ssl/certs/3fb36b73.0
  etc/ssl/certs/40193066.0
  etc/ssl/certs/4042bcee.0
  etc/ssl/certs/40547a79.0
  etc/ssl/certs/406c9bb1.0
  etc/ssl/certs/48bec511.0
  etc/ssl/certs/4b718d9b.0
  etc/ssl/certs/4bfab552.0
  etc/ssl/certs/4f316efb.0
  etc/ssl/certs/4fd49c6c.0
  etc/ssl/certs/5443e9e3.0
  etc/ssl/certs/54657681.0
  etc/ssl/certs/5860aaa6.0
  etc/ssl/certs/5931b5bc.0
  etc/ssl/certs/5ad8a5d6.0
  etc/ssl/certs/5cd81ad7.0
  etc/ssl/certs/5f15c80c.0
  etc/ssl/certs/5f618aec.0
  etc/ssl/certs/607986c7.0
  etc/ssl/certs/616816f6.0
  etc/ssl/certs/626dceaf.0
  etc/ssl/certs/653b494a.0
  etc/ssl/certs/68dd7389.0
  etc/ssl/certs/6a9bdba3.0
  etc/ssl/certs/6b99d060.0
  etc/ssl/certs/6d41d539.0
  etc/ssl/certs/6fa5da56.0
  etc/ssl/certs/706f604c.0
  etc/ssl/certs/749e9e03.0
  etc/ssl/certs/75d1b2ed.0
  etc/ssl/certs/76faf6c0.0
  etc/ssl/certs/7719f463.0
  etc/ssl/certs/773e07ad.0
  etc/ssl/certs/7a3adc42.0
  etc/ssl/certs/7a780d93.0
  etc/ssl/certs/7f3d5d1d.0
  etc/ssl/certs/7fa05551.0
  etc/ssl/certs/8160b96c.0
  etc/ssl/certs/81f2d2b1.0
  etc/ssl/certs/8312c4c1.0
  etc/ssl/certs/8508e720.0
  etc/ssl/certs/865fbdf9.0
  etc/ssl/certs/878d9bca.0
  etc/ssl/certs/8cb5ee0f.0
  etc/ssl/certs/8d86cdd1.0
  etc/ssl/certs/8d89cda1.0
  etc/ssl/certs/8f103249.0
  etc/ssl/certs/9046744a.0
  etc/ssl/certs/90c5a3c8.0
  etc/ssl/certs/930ac5d2.0
  etc/ssl/certs/93bc0acc.0
  etc/ssl/certs/9482e63a.0
  etc/ssl/certs/9846683b.0
  etc/ssl/certs/988a38cb.0
  etc/ssl/certs/9b46e03d.0
  etc/ssl/certs/9b5697b0.0
  etc/ssl/certs/9bf03295.0
  etc/ssl/certs/9c8dfbd4.0
  etc/ssl/certs/9d04f354.0
  etc/ssl/certs/9ef4a08a.0
  etc/ssl/certs/9f727ac7.0
  etc/ssl/certs/ACCVRAIZ1.crt
  etc/ssl/certs/AC_RAIZ_FNMT-RCM.crt
  etc/ssl/certs/AC_RAIZ_FNMT-RCM_SERVIDORES_SEGUROS.crt
  etc/ssl/certs/ANF_Secure_Server_Root_CA.crt
  etc/ssl/certs/Actalis_Authentication_Root_CA.crt
  etc/ssl/certs/AffirmTrust_Commercial.crt
  etc/ssl/certs/AffirmTrust_Networking.crt
  etc/ssl/certs/AffirmTrust_Premium.crt
  etc/ssl/certs/AffirmTrust_Premium_ECC.crt
  etc/ssl/certs/Amazon_Root_CA_1.crt
  etc/ssl/certs/Amazon_Root_CA_2.crt
  etc/ssl/certs/Amazon_Root_CA_3.crt
  etc/ssl/certs/Amazon_Root_CA_4.crt
  etc/ssl/certs/Atos_TrustedRoot_2011.crt
  etc/ssl/certs/Atos_TrustedRoot_Root_CA_ECC_TLS_2021.crt
  etc/ssl/certs/Atos_TrustedRoot_Root_CA_RSA_TLS_2021.crt
  etc/ssl/certs/Autoridad_de_Certificacion_Firmaprofesional_CIF_A62634068.crt
  etc/ssl/certs/BJCA_Global_Root_CA1.crt
  etc/ssl/certs/BJCA_Global_Root_CA2.crt
  etc/ssl/certs/Baltimore_CyberTrust_Root.crt
  etc/ssl/certs/Buypass_Class_2_Root_CA.crt
  etc/ssl/certs/Buypass_Class_3_Root_CA.crt
  etc/ssl/certs/CA_Disig_Root_R2.crt
  etc/ssl/certs/CFCA_EV_ROOT.crt
  etc/ssl/certs/COMODO_Certification_Authority.crt
  etc/ssl/certs/COMODO_ECC_Certification_Authority.crt
  etc/ssl/certs/COMODO_RSA_Certification_Authority.crt
  etc/ssl/certs/Certainly_Root_E1.crt
  etc/ssl/certs/Certainly_Root_R1.crt
  etc/ssl/certs/Certigna.crt
  etc/ssl/certs/Certigna_Root_CA.crt
  etc/ssl/certs/Certum_EC-384_CA.crt
  etc/ssl/certs/Certum_Trusted_Network_CA.crt
  etc/ssl/certs/Certum_Trusted_Network_CA_2.crt
  etc/ssl/certs/Certum_Trusted_Root_CA.crt
  etc/ssl/certs/CommScope_Public_Trust_ECC_Root-01.crt
  etc/ssl/certs/CommScope_Public_Trust_ECC_Root-02.crt
  etc/ssl/certs/CommScope_Public_Trust_RSA_Root-01.crt
  etc/ssl/certs/CommScope_Public_Trust_RSA_Root-02.crt
  etc/ssl/certs/Comodo_AAA_Services_root.crt
  etc/ssl/certs/D-TRUST_BR_Root_CA_1_2020.crt
  etc/ssl/certs/D-TRUST_BR_Root_CA_2_2023.crt
  etc/ssl/certs/D-TRUST_EV_Root_CA_1_2020.crt
  etc/ssl/certs/D-TRUST_EV_Root_CA_2_2023.crt
  etc/ssl/certs/D-TRUST_Root_Class_3_CA_2_2009.crt
  etc/ssl/certs/D-TRUST_Root_Class_3_CA_2_EV_2009.crt
  etc/ssl/certs/DigiCert_Assured_ID_Root_CA.crt
  etc/ssl/certs/DigiCert_Assured_ID_Root_G2.crt
  etc/ssl/certs/DigiCert_Assured_ID_Root_G3.crt
  etc/ssl/certs/DigiCert_Global_Root_CA.crt
  etc/ssl/certs/DigiCert_Global_Root_G2.crt
  etc/ssl/certs/DigiCert_Global_Root_G3.crt
  etc/ssl/certs/DigiCert_High_Assurance_EV_Root_CA.crt
  etc/ssl/certs/DigiCert_TLS_ECC_P384_Root_G5.crt
  etc/ssl/certs/DigiCert_TLS_RSA4096_Root_G5.crt
  etc/ssl/certs/DigiCert_Trusted_Root_G4.crt
  etc/ssl/certs/Entrust.net_Premium_2048_Secure_Server_CA.crt
  etc/ssl/certs/Entrust_Root_Certification_Authority.crt
  etc/ssl/certs/Entrust_Root_Certification_Authority_-_EC1.crt
  etc/ssl/certs/Entrust_Root_Certification_Authority_-_G2.crt
  etc/ssl/certs/FIRMAPROFESIONAL_CA_ROOT-A_WEB.crt
  etc/ssl/certs/GDCA_TrustAUTH_R5_ROOT.crt
  etc/ssl/certs/GLOBALTRUST_2020.crt
  etc/ssl/certs/GTS_Root_R1.crt
  etc/ssl/certs/GTS_Root_R2.crt
  etc/ssl/certs/GTS_Root_R3.crt
  etc/ssl/certs/GTS_Root_R4.crt
  etc/ssl/certs/GlobalSign_ECC_Root_CA_-_R4.crt
  etc/ssl/certs/GlobalSign_ECC_Root_CA_-_R5.crt
  etc/ssl/certs/GlobalSign_Root_CA.crt
  etc/ssl/certs/GlobalSign_Root_CA_-_R3.crt
  etc/ssl/certs/GlobalSign_Root_CA_-_R6.crt
  etc/ssl/certs/GlobalSign_Root_E46.crt
  etc/ssl/certs/GlobalSign_Root_R46.crt
  etc/ssl/certs/Go_Daddy_Class_2_CA.crt
  etc/ssl/certs/Go_Daddy_Root_Certificate_Authority_-_G2.crt
  etc/ssl/certs/HARICA_TLS_ECC_Root_CA_2021.crt
  etc/ssl/certs/HARICA_TLS_RSA_Root_CA_2021.crt
  etc/ssl/certs/Hellenic_Academic_and_Research_Institutions_ECC_RootCA_2015.crt
  etc/ssl/certs/Hellenic_Academic_and_Research_Institutions_RootCA_2015.crt
  etc/ssl/certs/HiPKI_Root_CA_-_G1.crt
  etc/ssl/certs/Hongkong_Post_Root_CA_3.crt
  etc/ssl/certs/ISRG_Root_X1.crt
  etc/ssl/certs/ISRG_Root_X2.crt
  etc/ssl/certs/IdenTrust_Commercial_Root_CA_1.crt
  etc/ssl/certs/IdenTrust_Public_Sector_Root_CA_1.crt
  etc/ssl/certs/Izenpe.com.crt
  etc/ssl/certs/Microsec_e-Szigno_Root_CA_2009.crt
  etc/ssl/certs/Microsoft_ECC_Root_Certificate_Authority_2017.crt
  etc/ssl/certs/Microsoft_RSA_Root_Certificate_Authority_2017.crt
  etc/ssl/certs/NAVER_Global_Root_Certification_Authority.crt
  etc/ssl/certs/NetLock_Arany_=Class_Gold=_Főtanúsítvány.crt
  etc/ssl/certs/OISTE_WISeKey_Global_Root_GB_CA.crt
  etc/ssl/certs/OISTE_WISeKey_Global_Root_GC_CA.crt
  etc/ssl/certs/QuoVadis_Root_CA_1_G3.crt
  etc/ssl/certs/QuoVadis_Root_CA_2.crt
  etc/ssl/certs/QuoVadis_Root_CA_2_G3.crt
  etc/ssl/certs/QuoVadis_Root_CA_3.crt
  etc/ssl/certs/QuoVadis_Root_CA_3_G3.crt
  etc/ssl/certs/SSL.com_EV_Root_Certification_Authority_ECC.crt
  etc/ssl/certs/SSL.com_EV_Root_Certification_Authority_RSA_R2.crt
  etc/ssl/certs/SSL.com_Root_Certification_Authority_ECC.crt
  etc/ssl/certs/SSL.com_Root_Certification_Authority_RSA.crt
  etc/ssl/certs/SSL.com_TLS_ECC_Root_CA_2022.crt
  etc/ssl/certs/SSL.com_TLS_RSA_Root_CA_2022.crt
  etc/ssl/certs/SZAFIR_ROOT_CA2.crt
  etc/ssl/certs/Sectigo_Public_Server_Authentication_Root_E46.crt
  etc/ssl/certs/Sectigo_Public_Server_Authentication_Root_R46.crt
  etc/ssl/certs/SecureSign_Root_CA12.crt
  etc/ssl/certs/SecureSign_Root_CA14.crt
  etc/ssl/certs/SecureSign_Root_CA15.crt
  etc/ssl/certs/SecureTrust_CA.crt
  etc/ssl/certs/Secure_Global_CA.crt
  etc/ssl/certs/Security_Communication_ECC_RootCA1.crt
  etc/ssl/certs/Security_Communication_RootCA2.crt
  etc/ssl/certs/Starfield_Class_2_CA.crt
  etc/ssl/certs/Starfield_Root_Certificate_Authority_-_G2.crt
  etc/ssl/certs/Starfield_Services_Root_Certificate_Authority_-_G2.crt
  etc/ssl/certs/SwissSign_Gold_CA_-_G2.crt
  etc/ssl/certs/T-TeleSec_GlobalRoot_Class_2.crt
  etc/ssl/certs/T-TeleSec_GlobalRoot_Class_3.crt
  etc/ssl/certs/TUBITAK_Kamu_SM_SSL_Kok_Sertifikasi_-_Surum_1.crt
  etc/ssl/certs/TWCA_CYBER_Root_CA.crt
  etc/ssl/certs/TWCA_Global_Root_CA.crt
  etc/ssl/certs/TWCA_Root_Certification_Authority.crt
  etc/ssl/certs/Telekom_Security_TLS_ECC_Root_2020.crt
  etc/ssl/certs/Telekom_Security_TLS_RSA_Root_2023.crt
  etc/ssl/certs/TeliaSonera_Root_CA_v1.crt
  etc/ssl/certs/Telia_Root_CA_v2.crt
  etc/ssl/certs/TrustAsia_Global_Root_CA_G3.crt
  etc/ssl/certs/TrustAsia_Global_Root_CA_G4.crt
  etc/ssl/certs/Trustwave_Global_Certification_Authority.crt
  etc/ssl/certs/Trustwave_Global_ECC_P256_Certification_Authority.crt
  etc/ssl/certs/Trustwave_Global_ECC_P384_Certification_Authority.crt
  etc/ssl/certs/TunTrust_Root_CA.crt
  etc/ssl/certs/UCA_Extended_Validation_Root.crt
  etc/ssl/certs/UCA_Global_G2_Root.crt
  etc/ssl/certs/USERTrust_ECC_Certification_Authority.crt
  etc/ssl/certs/USERTrust_RSA_Certification_Authority.crt
  etc/ssl/certs/XRamp_Global_CA_Root.crt
  etc/ssl/certs/a09a51ae.0
  etc/ssl/certs/a3418fda.0
  etc/ssl/certs/a89d74c2.0
  etc/ssl/certs/a94d09e5.0
  etc/ssl/certs/aee5f10d.0
  etc/ssl/certs/b0e59380.0
  etc/ssl/certs/b1159c4c.0
  etc/ssl/certs/b433981b.0
  etc/ssl/certs/b66938e9.0
  etc/ssl/certs/b727005e.0
  etc/ssl/certs/b7a5b843.0
  etc/ssl/certs/b81b93f0.0
  etc/ssl/certs/b8d25de6.0
  etc/ssl/certs/ba8887ce.0
  etc/ssl/certs/bf53fb88.0
  etc/ssl/certs/c01eb047.0
  etc/ssl/certs/c28a8a30.0
  etc/ssl/certs/ca6e4ad9.0
  etc/ssl/certs/cbf06781.0
  etc/ssl/certs/cc450945.0
  etc/ssl/certs/cd58d51e.0
  etc/ssl/certs/cd8c0d63.0
  etc/ssl/certs/ce5e74ef.0
  etc/ssl/certs/certSIGN_ROOT_CA.crt
  etc/ssl/certs/certSIGN_Root_CA_G2.crt
  etc/ssl/certs/d4dae3dd.0
  etc/ssl/certs/d52c538d.0
  etc/ssl/certs/d6325660.0
  etc/ssl/certs/d7e8dc79.0
  etc/ssl/certs/d887a5bb.0
  etc/ssl/certs/da0cfd1d.0
  etc/ssl/certs/dc4d6a89.0
  etc/ssl/certs/dd8e9d41.0
  etc/ssl/certs/ddcda989.0
  etc/ssl/certs/de6d66f3.0
  etc/ssl/certs/e-Szigno_Root_CA_2017.crt
  etc/ssl/certs/e113c810.0
  etc/ssl/certs/e18bfb83.0
  etc/ssl/certs/e35234b1.0
  etc/ssl/certs/e36a6752.0
  etc/ssl/certs/e73d606e.0
  etc/ssl/certs/e868b802.0
  etc/ssl/certs/e8de2f56.0
  etc/ssl/certs/ePKI_Root_Certification_Authority.crt
  etc/ssl/certs/ecccd8db.0
  etc/ssl/certs/ed858448.0
  etc/ssl/certs/ee64a828.0
  etc/ssl/certs/eed8c118.0
  etc/ssl/certs/ef954a4e.0
  etc/ssl/certs/emSign_ECC_Root_CA_-_C3.crt
  etc/ssl/certs/emSign_ECC_Root_CA_-_G3.crt
  etc/ssl/certs/emSign_Root_CA_-_C1.crt
  etc/ssl/certs/emSign_Root_CA_-_G1.crt
  etc/ssl/certs/f081611a.0
  etc/ssl/certs/f0c70a8d.0
  etc/ssl/certs/f249de83.0
  etc/ssl/certs/f30dd6ad.0
  etc/ssl/certs/f387163d.0
  etc/ssl/certs/f39fc864.0
  etc/ssl/certs/f51bb24c.0
  etc/ssl/certs/fa5da96b.0
  etc/ssl/certs/fb717492.0
  etc/ssl/certs/fc5a8f99.0
  etc/ssl/certs/fd64f3fc.0
  etc/ssl/certs/fe8a2cd8.0
  etc/ssl/certs/feffd413.0
  etc/ssl/certs/ff34af3f.0
  etc/ssl/certs/ffdd40f9.0
  etc/ssl/certs/vTrus_ECC_Root_CA.crt
  etc/ssl/certs/vTrus_Root_CA.crt
  lib/apk/packages/ca-certificates.list
```

And again, I'm not in favor of overloading/abusing unrelated funcitonality.

Also, this PR is not just about ca-certificates, but adding support for virtual provides.

map-b commented:
> build: fix apk packaging and ABI-versioning
> The updated logic for the APK dependencies and provides is as follows:
> 
> - If ABI version is defined:
>   - package is named `package_name-ABI_version`
>   - package implicitly provides
>     `package_name-ABI_version=package_version`
>     this implies that only one version of a package per ABI can be
>     installed at the same time
>   - additionally provide `package_name` so multiple packages can be
>     looked up by its base name
>   - for each `provides`, provide `provide-ABI_version=package_version`
>     this implies that only one version of a provide can be installed at
>     the same time
> 
> - else if ABI version is _not_ defined
>   - package is named `package_name`
>   - package implicitly provides `package_name=package_version`
>     this implies that only one version of a package can be installed at
>     the same time
>   - if `alternatives` is defined
>     - for each `provides`, provide `provide`
>       this implies that multiple versions of a provide can be installed
>       at the same time
>   - else if `alternatives` is _not_ defined
>     - for each `provides`, provide `provide=package_version`
>       this implies that only one version of a provide can be installed
>       at the same time
> 
> Both cases a package can be looked up by its base name.
> 

The only case we allow side by side packages installations is for packages using the "alternatives" mechanism and in most cases it will fit what you expect of symlinking binaries defining alternative priorities and some cases like this one maybe it will not but the concept would be still the same and I do not think this would be that "creative" overload of this mechanism.

Looking at openwrt and packages repos, there are not a huge list of cases using "provides" and the most use the alternative mechanism.

In general it seems that if several packages have a common provides and are meant to installed side-by-side those will need to implement the "alternatives" mechanism. Excuse me if I did not articulate my thoughts better before.

nxhack commented:
After applying this commit, I performed a clean build, but this time a similar error occurs related to 'wget' package.
```
ERROR: unable to select packages:
  uclient-fetch-2025.10.03~dc909ca7-r1:
    conflicts: wget-ssl-1.25.0-r1[wget=2025.10.03~dc909ca7-r1]
    satisfies: world[uclient-fetch] apk-openssl-3.0.2-r1[wget]
  wget-ssl-1.25.0-r1:
    conflicts: uclient-fetch-2025.10.03~dc909ca7-r1[wget]
    satisfies: world[wget-ssl] apk-openssl-3.0.2-r1[wget]
```

GeorgeSapkin commented:
@schuettecarsten I'm going to look into apk/wget today, but aren't the iptables ones as expected? I thought nft and legacy were conflicts.

```
  ip6tables-nft-1.8.10-r2:
    conflicts: ip6tables-zz-legacy-1.8.10-r2[ip6tables=1.8.10-r2]
    satisfies: world[ip6tables-nft] dockerd-28.5.1-r1[ip6tables]
  iptables-nft-1.8.10-r2:
    conflicts: iptables-zz-legacy-1.8.10-r2[iptables=1.8.10-r2]
    satisfies: world[iptables-nft] dockerd-28.5.1-r1[iptables] sqm-scripts-1.6.0-r1[iptables]
  ip6tables-zz-legacy-1.8.10-r2:
    conflicts: ip6tables-nft-1.8.10-r2[ip6tables=1.8.10-r2]
    satisfies: world[ip6tables-zz-legacy] dockerd-28.5.1-r1[ip6tables]
  iptables-zz-legacy-1.8.10-r2:
    conflicts: iptables-nft-1.8.10-r2[iptables=1.8.10-r2]
    satisfies: world[iptables-zz-legacy] dockerd-28.5.1-r1[iptables] sqm-scripts-1.6.0-r1[iptables]
```

GeorgeSapkin commented:
> This way we don't need to rely on packages setting any conflicts/provides/whatever fields correctly to prevent parallel installation, we just automatically catch it at install time.

The use case for provides including the virtual ones is for other packages to depend on them, and not just conflict resolution. Other than that I agree. And again, think that packages should define these restrictions themselves.

teleostnacl commented:
When apply this commit, executing `make`, the following warning will be reported:
```text
WARNING: Makefile 'package/feeds/packages/nmap/Makefile' has a dependency on 'ca-certs', which does not exist
WARNING: Makefile 'package/feeds/packages/nmap/Makefile' has a dependency on 'ca-certs', which does not exist
WARNING: Makefile 'package/feeds/packages/nmap/Makefile' has a dependency on 'ca-certs', which does not exist
WARNING: Makefile 'package/feeds/packages/nmap/Makefile' has a dependency on 'ca-certs', which does not exist
WARNING: Makefile 'package/feeds/packages/nmap/Makefile' has a dependency on 'ca-certs', which does not exist
WARNING: Makefile 'package/feeds/packages/onionshare-cli/Makefile' has a dependency on 'python3-pysocks', which does not exist
WARNING: Makefile 'package/feeds/packages/onionshare-cli/Makefile' has a dependency on 'python3-unidecode', which does not exist
WARNING: Makefile 'package/feeds/packages/python3/Makefile' has a dependency on 'ca-certs', which does not exist
WARNING: Makefile 'package/feeds/packages/stubby/Makefile' has a dependency on 'ca-certs', which does not exist
```

GeorgeSapkin commented:
@teleostnacl yeah, I'm looking into package-metadata as well.

GeorgeSapkin commented:
> Do I need to manually add `@` to all the `PROVIDES` ？

There's no blanket answer, as every package is doing something unique. As the first step you need to look at the packages first, and understand if they are even supposed to work side-by-side or are they conflicting variants.

map-b commented:
@KanjiMonster Discussion about file conflicts with APK https://github.com/openwrt/openwrt/issues/20802#issuecomment-3539469449 the "origin" field has hard implications, I was thinking about appending the "variant" to origin to avoid that, at least, in those cases but I thought, like you've just said, we would want always to not allow silently file overwrites. (APK has another field, like "replaces" which explicitly allow that if needed), we can reopen that issue

GeorgeSapkin commented:
I fixed up the alternatives logic and uclient-fetch and wget should install side-by-side now (once rebuilt of course):

```bash
root@OpenWrt:/tmp# apk query --installed --fields name,origin,depends,provides uclient-fetch wget-ssl
Name: uclient-fetch
Origin: /home/george/Dev/openwrt-openwrt/package/libs/uclient
Depends: libc libuclient20201210
Provides: wget

Name: wget-ssl
Origin: /home/george/Dev/openwrt-packages/net/wget
Depends: libc libopenssl3 libpcre2 librt zlib
Provides: gnu-wget wget
```

@nxhack @schuettecarsten @teleostnacl can you check the above scenario?

But the more I look at it the more I dislike the current provides/alternatives situation. I'm sure there's a package that breaks this.

I'll look at the kmod thing next, unless @efahl has some input.

efahl commented:
Oh, to clarify that first note: the apk solver never even looks at the `provider_priority` since the version number check in `provides: kmod-r8169=blah` makes the decision before it gets to that tie-breaker.

GeorgeSapkin commented:
Looks like lots of package specify `PROVIDES:=self`, where `self` is the name of a package, which apk doesn't like (and it doesn't make sense either way), so I'm going to add a filter to remove it. But realistically it shouldn't be in package specs.

@efahl shall I clean-up and add the kmod fix as separate commit? This kind of expands the scope of the PR, but so be it.

GeorgeSapkin commented:
Can kmods have an ABI version? :thinking: I see it used for cfg80211 only, but judging from blame it's a kludge to fix triggering dependency building.

efahl commented:
> so I'm going to add a filter to remove it. But realistically it shouldn't be in package specs.

Sounds like a plan, simply ignore adding provides when package-name == provides-name, and maybe later make it a warning.  Then package maintainers can start picking away at the packages with the bad definitions.

nxhack commented:
@GeorgeSapkin 
> I fixed up the alternatives logic and uclient-fetch and wget should install side-by-side now (once rebuilt of course):

~~“package/install” succeeded and the build is complete. However, the target image does not have the “wget” package installed for the configuration being installed.~~

edit: I'll verify it a little more.

KanjiMonster commented:
> Looks like lots of package specify `PROVIDES:=self`, where `self` is the name of a package, which apk doesn't like (and it doesn't make sense either way), so I'm going to add a filter to remove it. But realistically it shouldn't be in package specs.

Be aware that the package specs including the `PROVIDES` is also used for Kconfig (`tmp/.config-package.in`) generation. IIRC this is the reason some packages have a provides for themselves.

map-b commented:
More read the APK docs more convince myself that the virtual package APK concept is more general and required for us than we have been thought, the only way to explicitly set priorities in providers is not using versioned provides so almost every use of provides in openwrt will need to be an apk virtual package name.

The "kmod-" case is making it clear, we also need a more flexible way to define those priorities (you know what I am thinking about those alternatives priorities).

```
*provider-priority* (*provider_priority*, *k*)
	This determines the default installation priority for the non-versioned
	package names the packages lists in the *provides* field. By default
	a non-versioned provides will not be selected automatically for installation.
	But specifying *provider-priority* enables this automatic selection, and is
	used to determine which of the packages to install in case multiple packages
	provide the same non-versioned package name.

*depends* (*depend*, *D*)
	List of dependencies for the package. Installing this package will
	require APK to first satisfy the list of all its dependencies.

	The dependencies are used by various APK components:
	- The solver will try to find a solution that all package dependencies
	  are satisfied (as well as the world dependencies)
	- When apk is committing changes to the file system, it will install
	  or remove packages in such order that all dependencies of the package
	  will be satisfied (assuming there are no circular dependencies)
	- When apk runs the package trigger scripts, they will be ordered
	  so that the triggers of all dependencies before running the trigger
	  for this package

*provides* (*provides*, *p*)
	List of package names (and optionally its version) this package
	provides in addition to its primary name and version. The provided
	name can contain additionally characters: comma (,), brackets ([]), 	colons (:) and slashes (/) in the name.	This allows using namespaces 	for automatically generated names. 

	If the provided name contains a version number:
	- the solver will treat it as-if a real package with the provided
	  name is installed
	- the package becomes automatically selectable by anything depending
	  on the provided name
	- the package will automatically become the single possible owner
	  for the provided name
	- the package will automatically conflict with any package with
	  the same primary or provided package name

	If the provided name does not include version:
	- the package is not automatically selectable for installation
	  by that fact that there is a dependency on the provided name
		- specifying *provides_priority* will allow automatic selection
		- otherwise user is expected to manually select one of the
		  concrete package names in world which allows selection
	- the package is not considered to own provided name
	- multiple packages provided the same name without a version are
	  allowed to be installed simultaneously
	- apk internally considers a package name with only non-versioned
	  providers as a "virtual package name"
```

https://gitlab.alpinelinux.org/alpine/apk-tools/-/blob/master/doc/apk-package.5.scd?#L166

GeorgeSapkin committed:
build: provide virtual self in kmods

Add an implicit self-provide to kmods. apk can't handle self provides,
be it versioned or virtual, so opt for a prefix and a suffix instead.
Package name without a prefix/suffix is too generic and might conflict
with other packages, e.g. wireguard. This allows several variants to
provide the same virtual package without adding extra provides to the
default one, e.g. r8169 implicitly provides kmod-r8169-any and is marked
as default, so r8125 can explicitly provide @kmod-r8169-any as well.

Signed-off-by: George Sapkin <george@sapk.in>
Link: https://github.com/openwrt/openwrt/pull/21288
Signed-off-by: Robert Marko <robimarko@gmail.com>

GeorgeSapkin committed:
kernel/r8169: provide a virtual kmod

Switch r8169 and related kmods to use the new virtual kmod provides
semantic and mark r8169 as the default variant.

Signed-off-by: George Sapkin <george@sapk.in>
Link: https://github.com/openwrt/openwrt/pull/21288
Signed-off-by: Robert Marko <robimarko@gmail.com>

GeorgeSapkin committed:
kernel/ath10k: provide a virtual kmod

Switch ath10k and related kmods to use the new virtual kmod provides
semantic and mark ath10k as the default variant.

Signed-off-by: George Sapkin <george@sapk.in>
Link: https://github.com/openwrt/openwrt/pull/21288
Signed-off-by: Robert Marko <robimarko@gmail.com>

GeorgeSapkin committed:
kernel/rtl8812au-ct: provide a virtual kmod

Switch rtl8812au-ct to use the new virtual kmod provides semantic.

Signed-off-by: George Sapkin <george@sapk.in>
Link: https://github.com/openwrt/openwrt/pull/21288
Signed-off-by: Robert Marko <robimarko@gmail.com>

GeorgeSapkin commented:
@efahl it took me a minute to get a fresh buildroot going today, but here's the latest take on the kmod provides with a lot less packaging magic. I added a virtual self provider on all kmods, added a default variant to kmods with multiple versions, and kmods that had a provide (luckily just a few) provide the same virtual provider as the default ones.

Here's an example with ath10k and ath10k-ct:

```bash
staging_dir/host/bin/apk adbdump ./bin/targets/x86/64/packages/kmod-ath10k-6* | head -20 | tail -6
  provider-priority: 100
  depends: # 2 items
    - kernel=6.12.62~2b02ee8dd9577e502325f4a36898816b-r1
    - kmod-ath
  provides: # 1 items
    - kmod-ath10k

staging_dir/host/bin/apk adbdump ./bin/targets/x86/64/packages/kmod-ath10k-ct* | head -21 | tail -9
  provider-priority: 1
  depends: # 4 items
    - kernel=6.12.62~2b02ee8dd9577e502325f4a36898816b-r1
    - kmod-ath
    - kmod-hwmon-core
    - kmod-mac80211
  provides: # 2 items
    - kmod-ath10k
    - kmod-ath10k-ct
```

And here's one with r8169 and r8125 (I'm not sure I'm happy having these memorized):

```bash
staging_dir/host/bin/apk adbdump ./bin/targets/x86/64/packages/kmod-r8169* | head -20 | tail -9
  provider-priority: 100
  depends: # 5 items
    - kernel=6.12.62~2b02ee8dd9577e502325f4a36898816b-r1
    - kmod-mdio-devres
    - kmod-mii
    - kmod-phy-realtek
    - r8169-firmware
  provides: # 1 items
    - kmod-r8169

staging_dir/host/bin/apk adbdump ./bin/targets/x86/64/packages/kmod-r8125* | head -19 | tail -7
  provider-priority: 1
  depends: # 2 items
    - kernel=6.12.62~2b02ee8dd9577e502325f4a36898816b-r1
    - kmod-libphy
  provides: # 2 items
    - kmod-r8125
    - kmod-r8169
```

I find this solution a lot less cumbersome and it reuses the same new semantic from earlier.

I haven't done much testing so far with the kmod part.

efahl commented:
> As an aside, would be useful, if apk would refuse to pack something it won't install 🤔

Yeah, might warrant an issue over on apk-tools asking if this could be an enhancement, they're usually quite receptive to requests like that.

efahl commented:
> But again that r8169 situation is confusing.

Seems to be working as I expect when I use the default, add a package, remove the default...
```bash
# Default, should have just r8169
$ make manifest  | grep kmod-r8
.../apk add --arch x86_64 --no-scripts apk-mbedtls ... kmod-r8169 ...
  kmod-r8169 (6.12.62-r1)
(127/149) Installing kmod-r8169 (6.12.62-r1)
lib/apk/packages/kmod-r8169.list
kmod-r8169 - 6.12.62-r1

# Add r8127, should show both r8169 and r8127
$ make manifest PACKAGES=kmod-r8127  | grep kmod-r8
.../apk add --arch x86_64 --no-scripts apk-mbedtls ... kmod-r8127 ... kmod-r8169 ...
  kmod-r8127 (6.12.62.11.015.00-r1)
  kmod-r8169 (6.12.62-r1)
(124/150) Installing kmod-r8127 (6.12.62.11.015.00-r1)
lib/apk/packages/kmod-r8127.list
(128/150) Installing kmod-r8169 (6.12.62-r1)
lib/apk/packages/kmod-r8169.list
kmod-r8127 - 6.12.62.11.015.00-r1
kmod-r8169 - 6.12.62-r1

# Add 8127 and remove 8169, should show just r8127
$ make manifest PACKAGES="-kmod-r8169 kmod-r8127"  | grep kmod-r8
.../apk add --arch x86_64 --no-scripts apk-mbedtls ... kmod-r8127 ...
  kmod-r8127 (6.12.62.11.015.00-r1)
(124/146) Installing kmod-r8127 (6.12.62.11.015.00-r1)
lib/apk/packages/kmod-r8127.list
kmod-r8127 - 6.12.62.11.015.00-r1

# Just remove the default, should show nothing
$ make manifest PACKAGES="-kmod-r8169"  | grep kmod-r8
<nothing>
```

GeorgeSapkin commented:
Sweet. I'm going play around with a more readable format. Something like: `@kmod-X-any`. Guess where the inspiration is coming from :sweat_smile:

Might as well make all packages provide `@X-any`, so it's easier to define variants without creating custom virtual provides each time, like `ca-certs` for `ca-certificates`.

map-b commented:
```
        $(FAKEROOT) $(STAGING_DIR_HOST)/bin/apk mkpkg \
          --info "name:$(1)$$(ABIV_$(1))" \
          --info "version:$(VERSION)" \
```
We are literally fighting against implicit APK behavior of providing the versioned  name `package_name=package_version`, since it does not allow to make that name virtual, should we do not pass the "version" at first and only add the version in the "provides" field when needed.

The function of these two ways of define a version should be equivalent except for the case APK info commands showing a empty version field? 

At least for kmod- cases?

(Taking for granted that not giving a version to `apk mkapk` `--version` field would make "package_name" a virtual one)

map-b commented:
We are packaging all packages with versioned names first and then declaring the provides as virtual, I was thinking a way to do the opposite: start with virtual names and then version them in the provides when needed.(If that is even possible) (This would allow self-provides?? for example)

Do take me bad, just thoughts trying to find the best path, amazing job anyway.

efahl commented:
This thing is stress testing the living hell out of those CI check builds...

efahl commented:
@robimarko All checks passed, my testing looks good, time for your review.  😁

robimarko commented:
Thanks! Rebased on top of main and merged!

Djfe commented:
does this mean ath10k-ct is not the default anymore?

teleostnacl commented:
I'm sorry, but I meet another error:
```
ERROR: unable to select packages:
  kernel-6.12.63~a64cdc69add322a87084e41c2857026c-r1:
    breaks: kmod-nf-conntrack-netlink-6.12.62-r1[kernel=6.12.62~cdbbaf6403094fd68db5b0a1ec5c9fd8-r1]
    satisfies: world[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-cfg80211-6.12.63.6.18-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-crypto-aead-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-crypto-authenc-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-crypto-ccm-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-crypto-cmac-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-crypto-crc32c-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-crypto-ctr-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-crypto-des-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-crypto-gcm-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-crypto-geniv-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-crypto-gf128-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-crypto-ghash-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-crypto-hash-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-crypto-hmac-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-crypto-hw-safexcel-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-crypto-manager-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-crypto-md5-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-crypto-null-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-crypto-rng-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-crypto-seqiv-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-crypto-sha1-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-crypto-sha256-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-crypto-sha3-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-crypto-sha512-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-gpio-button-hotplug-6.12.63-r5[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-hwmon-core-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-i2c-core-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-leds-gpio-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-lib-crc-ccitt-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-lib-crc32c-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-mac80211-6.12.63.6.18-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-mt76-connac-6.12.63.2025.11.06~eb567bc7-r2[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-mt76-core-6.12.63.2025.11.06~eb567bc7-r2[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-mt7915e-6.12.63.2025.11.06~eb567bc7-r2[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-mt7981-firmware-6.12.63.2025.11.06~eb567bc7-r2[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-nf-conntrack-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-nf-conntrack6-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-nf-flow-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-nf-log-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-nf-log6-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-nf-nat-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-nf-reject-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-nf-reject6-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-nfnetlink-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-nft-core-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-nft-fib-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-nft-nat-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-nft-offload-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-nls-base-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-ppp-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-pppoe-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-pppox-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-slhc-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-thermal-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-usb-common-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-usb-core-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-usb-xhci-hcd-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-usb-xhci-mtk-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
               kmod-usb3-6.12.63-r1[kernel=6.12.63~a64cdc69add322a87084e41c2857026c-r1]
```

efahl commented:
Is this the correct pattern?

1) Add `DEFAULT_VARIANT` on base/preferred package.
2) Add/change-to `PROVIDES:=@dnsmasq-any` on all variants.

```
git diff package/network/services/dnsmasq/Makefile
diff --git a/package/network/services/dnsmasq/Makefile b/package/network/services/dnsmasq/Makefile
index 6b9853e5ff8e..b7a4a1b0ab10 100644
--- a/package/network/services/dnsmasq/Makefile
+++ b/package/network/services/dnsmasq/Makefile
@@ -51,6 +51,8 @@ endef
 define Package/dnsmasq
 $(call Package/dnsmasq/Default)
   VARIANT:=nodhcpv6
+  DEFAULT_VARIANT:=1
+  PROVIDES:=@dnsmasq-any
 endef

 define Package/dnsmasq-dhcpv6
@@ -58,7 +60,7 @@ $(call Package/dnsmasq/Default)
   TITLE += (with DHCPv6 support)
   DEPENDS+=@IPV6
   VARIANT:=dhcpv6
-  PROVIDES:=dnsmasq
+  PROVIDES:=@dnsmasq-any
 endef

 define Package/dnsmasq-full
@@ -69,7 +71,7 @@ $(call Package/dnsmasq/Default)
        +PACKAGE_dnsmasq_full_conntrack:libnetfilter-conntrack \
        +PACKAGE_dnsmasq_full_nftset:nftables-json
   VARIANT:=full
-  PROVIDES:=dnsmasq
+  PROVIDES:=@dnsmasq-any
 endef

 define Package/dnsmasq/description
```

GeorgeSapkin commented:
Interesting. Besides the unnecessary priority, everything _seems_ OK. I'll have a look.

GeorgeSapkin commented:
> Is this the correct pattern?

Yes and you can colorize your diffs by doing ` ```diff `, just in case:

```diff
diff --git a/package/network/services/dnsmasq/Makefile b/package/network/services/dnsmasq/Makefile
index 6b9853e5ff8e..b7a4a1b0ab10 100644
--- a/package/network/services/dnsmasq/Makefile
+++ b/package/network/services/dnsmasq/Makefile
@@ -51,6 +51,8 @@ endef
 define Package/dnsmasq
 $(call Package/dnsmasq/Default)
   VARIANT:=nodhcpv6
+  DEFAULT_VARIANT:=1
+  PROVIDES:=@dnsmasq-any
 endef

 define Package/dnsmasq-dhcpv6
@@ -58,7 +60,7 @@ $(call Package/dnsmasq/Default)
   TITLE += (with DHCPv6 support)
   DEPENDS+=@IPV6
   VARIANT:=dhcpv6
-  PROVIDES:=dnsmasq
+  PROVIDES:=@dnsmasq-any
 endef

 define Package/dnsmasq-full
@@ -69,7 +71,7 @@ $(call Package/dnsmasq/Default)
        +PACKAGE_dnsmasq_full_conntrack:libnetfilter-conntrack \
        +PACKAGE_dnsmasq_full_nftset:nftables-json
   VARIANT:=full
-  PROVIDES:=dnsmasq
+  PROVIDES:=@dnsmasq-any
 endef

 define Package/dnsmasq/description
```

I'm curious what's up with the priority though.

dioni21 commented:
@efahl Do you need any beta testers for your pre-releases?
How could I help?

efahl commented:
> I guess I'll have to dig into the solver at some point 🙈

That's always fun.  It's pretty convoluted, I haven't been able to trace through it by hand for any meaningful case, and am insufficiently motivated to set it up so I can chase through with a debugger.

In any case, what's our way forward?  Manually add `DEFAULT_VARIANT:=1` to all the problematic packages?  I can't see a way to detect when variants exist from within a specific variant in order to do any automated setting of priority.

I've been using `git grep CONFLICTS` to identify candidates, as that seems to be used across all the interesting packages (couple dozen in openwrt, maybe 50 in the packages repo, haven't looked in luci, telephony etc).

GeorgeSapkin commented:
dnsmasq debug log for posterity:

```
name <dnsmasq> selected from selectable list
select_package: dnsmasq (requirers=1, autosel=1, iif=0, order_id=0x4000005f)
  consider dnsmasq-2.91-r2 iif_triggered=0, tag_ok=1, selectable=1, available=1, flags=0x0, provider_priority=0, installed=0
   prefer existing package
    choose as new provider
  consider dnsmasq-dhcpv6-2.91-r2 iif_triggered=0, tag_ok=1, selectable=1, available=1, flags=0x0, provider_priority=1, installed=0
    prefer highest declared provider priority
    choose as new provider
  consider dnsmasq-full-2.91-r2 iif_triggered=0, tag_ok=1, selectable=1, available=1, flags=0x0, provider_priority=1, installed=0
    prefer lowest available repository
selecting: dnsmasq-dhcpv6-2.91-r2, available: 1
assign dnsmasq-dhcpv6 to dnsmasq-dhcpv6-2.91-r2
assign dnsmasq to dnsmasq-dhcpv6-2.91-r2
disqualify_package: dnsmasq-2.91-r2 (conflicting provides)
disqualify_package: dnsmasq-full-2.91-r2 (conflicting provides)
    apply_constraint: libc
    apply_constraint: provider: libc-1.2.5-r5: 1
```

So it really looks like it's just priority :shrug:

Will try setting the default and see what's in the log.

efahl commented:
Another case looks good:
```
$ make manifest PACKAGES="wpad-basic-mbedtls wpad-mesh-mbedtls" | grep wpad
... apk ... add --arch x86_64 --no-scripts wpad-basic-mbedtls wpad-mesh-mbedtls ...

ERROR: unable to select packages:
  wpad-basic-mbedtls-2025.08.26~ca266cc2-r1:
    conflicts: wpad-mesh-mbedtls-2025.08.26~ca266cc2-r1[hostapd=2025.08.26~ca266cc2-r1]
               wpad-mesh-mbedtls-2025.08.26~ca266cc2-r1[wpa-supplicant=2025.08.26~ca266cc2-r1]
    satisfies: world[wpad-basic-mbedtls]
  wpad-mesh-mbedtls-2025.08.26~ca266cc2-r1:
    conflicts: wpad-basic-mbedtls-2025.08.26~ca266cc2-r1[hostapd=2025.08.26~ca266cc2-r1]
               wpad-basic-mbedtls-2025.08.26~ca266cc2-r1[wpa-supplicant=2025.08.26~ca266cc2-r1]
    satisfies: world[wpad-mesh-mbedtls]
make[2]: *** [Makefile:227: package_install] Error 2
make[1]: *** [Makefile:167: _call_manifest] Error 2
make: *** [Makefile:345: manifest] Error 2
```

That was my last weird-case test, PR incoming...

GeorgeSapkin commented:
> Did you find where it sets the default for priority? I ran across one case where if it's missing, the solver uses 10

Perhaps you mixed it up with radix being 10? If it's not set it's 0 as in the log above for dnsmasq.

Djfe commented:
> I suspect that what @Djfe is seeing on ath10k is the same...

I was not seeing anything, I only read that one commit's message which changed ath10k and got confused. I haven't done any tests whatsoever 😅

GeorgeSapkin commented:
I adjusted the default priority to 0 and without a default variant set a correct dnsmasq variant is selected:

```
name <dnsmasq> selected from selectable list
select_package: dnsmasq (requirers=1, autosel=1, iif=0, order_id=0x4000005f)
  consider dnsmasq-2.91-r2 iif_triggered=0, tag_ok=1, selectable=1, available=1, flags=0x0, provider_priority=0, installed=0
   prefer existing package
    choose as new provider
  consider dnsmasq-dhcpv6-2.91-r2 iif_triggered=0, tag_ok=1, selectable=1, available=1, flags=0x0, provider_priority=0, installed=0
    prefer lowest available repository
  consider dnsmasq-full-2.91-r2 iif_triggered=0, tag_ok=1, selectable=1, available=1, flags=0x0, provider_priority=0, installed=0
    prefer lowest available repository
selecting: dnsmasq-2.91-r2, available: 1
assign dnsmasq to dnsmasq-2.91-r2
disqualify_package: dnsmasq-dhcpv6-2.91-r2 (conflicting provides)
disqualify_package: dnsmasq-full-2.91-r2 (conflicting provides)
    apply_constraint: libc
    apply_constraint: provider: libc-1.2.5-r5: 1
```

The theory is that the package with the actual matching name should win. I'll check the other packages too.

Edit: the explicit option is better of course, but I don't think this should silently fail.

efahl commented:
> Perhaps you mixed it up with radix being 10?

Oh, shit, I'll bet your right!  Taking a look, sure enough:
```
$ git grep 'priority.*10'
src/database.c:                 ipkg->replaces_priority = apk_blob_pull_uint(&l, 10);
src/database.c:         apk_blob_push_uint(&bbuf, ipkg->replaces_priority, 10);
src/database.c:         ipkg->replaces_priority = apk_blob_pull_uint(&r, 10);
src/package.c:          pkg->provider_priority = apk_blob_pull_uint(&value, 10);
src/package.c:          apk_blob_push_uint(&bbuf, info->provider_priority, 10);
```
and 
```
uint64_t apk_blob_pull_uint(apk_blob_t *b, int radix)
```

efahl commented:
Ok, so maybe we need to deal with this globally somehow here???

```
include/package-pack.mk:          $(if $(DEFAULT_VARIANT),--info "provider-priority:100",$(if $(PROVIDES),--info "provider-priority:1")) \
```

efahl commented:
I wonder if we've worked around @robimarko's setting it to '1'? https://github.com/openwrt/openwrt/commit/dea839773c27430031c2a59c3d37ebdfef3937b9

GeorgeSapkin commented:
Curiously it mentions the virtual part again (same as what I assumed from the docs), but I don't see anything like that in the solver. Could it be it's from before apk 3?

GeorgeSapkin commented:
Also I don't think that description is relevant, since the whole provides logic is completely different now to account for conflicts, etc.

Edit: and I think it's better to set the priority explicitly either way.

GeorgeSapkin commented:
I tested my theory that the matching name (i.e. `busybox` for `busybox`) is selected if priorities are equal based on the fact that the matching name lexicographically comes first in a repo (the final check in `compare_providers` in solver returns 0 and the first matching provider is selected, which happens to be the correct one). 

I reordered both definitions and eval/calls:

```Makefile
define Package/busybox-selinux
  ...
endef

define Package/busybox
  ...
endef

$(eval $(call BuildPackage,busybox-selinux))
$(eval $(call BuildPackage,busybox))
```

and apk still picks the right package:

```
name <busybox> selected from selectable list
select_package: busybox (requirers=3, autosel=1, iif=0, order_id=0x40000013)
  consider busybox-1.37.0-r6 iif_triggered=0, tag_ok=1, selectable=1, available=1, flags=0x0, provider_priority=0, installed=0
   prefer existing package
    choose as new provider
  consider busybox-selinux-1.37.0-r6 iif_triggered=0, tag_ok=1, selectable=1, available=1, flags=0x0, provider_priority=0, installed=0
    prefer lowest available repository
selecting: busybox-1.37.0-r6, available: 1
assign busybox to busybox-1.37.0-r6
disqualify_package: busybox-selinux-1.37.0-r6 (conflicting provides)
    apply_constraint: libc
    apply_constraint: provider: libc-1.2.5-r5: 1

select_package: dnsmasq (requirers=1, autosel=1, iif=0, order_id=0x40000064)
  consider dnsmasq-2.91-r2 iif_triggered=0, tag_ok=1, selectable=1, available=1, flags=0x0, provider_priority=0, installed=0
   prefer existing package
    choose as new provider
  consider dnsmasq-dhcpv6-2.91-r2 iif_triggered=0, tag_ok=1, selectable=1, available=1, flags=0x0, provider_priority=0, installed=0
    prefer lowest available repository
  consider dnsmasq-full-2.91-r2 iif_triggered=0, tag_ok=1, selectable=1, available=1, flags=0x0, provider_priority=0, installed=0
    prefer lowest available repository
selecting: dnsmasq-2.91-r2, available: 1
assign dnsmasq to dnsmasq-2.91-r2
disqualify_package: dnsmasq-dhcpv6-2.91-r2 (conflicting provides)
disqualify_package: dnsmasq-full-2.91-r2 (conflicting provides)
    apply_constraint: libc
    apply_constraint: provider: libc-1.2.5-r5: 1

name <procd> selected from selectable list
select_package: procd (requirers=3, autosel=1, iif=0, order_id=0x40000032)
  consider procd-2025.10.04~3b3501ab-r1 iif_triggered=0, tag_ok=1, selectable=1, available=1, flags=0x0, provider_priority=0, installed=0
   prefer existing package
    choose as new provider
  consider procd-selinux-2025.10.04~3b3501ab-r1 iif_triggered=0, tag_ok=1, selectable=1, available=1, flags=0x0, provider_priority=0, installed=0
    prefer lowest available repository
selecting: procd-2025.10.04~3b3501ab-r1, available: 1
assign procd to procd-2025.10.04~3b3501ab-r1
disqualify_package: procd-selinux-2025.10.04~3b3501ab-r1 (conflicting provides)
    apply_constraint: jshn
    apply_constraint: provider: jshn-2025.12.08~7928f171-r1: 1
```

I'm not sure how stable the package ordering is an repo or where it's coming from yet, but I'm still leaning towards using the apk defaults.

GeorgeSapkin commented:
Priority being set to 1 in #17008 sort of makes sense if there's only one virtual package, it's not marked as default and something depends on that. Since only a few packages are actually virtual now due to recent changes, I don't think it's necessary.

Speaking of which, lots provides that are not variants of the same package, but span multiple packages don't make sense anymore and are most likely working by chance, since a package with the highest version will be selected (the same thing that happened with out-of-kernel kmods and their version suffix). So this will need to be adjusted bit by bit.

dioni21 commented:
@efahl


    $ make manifest PACKAGES="dnsmasq" | grep dnsmasq
    < should show ONLY dnsmasq, no -full or -dhcpv6 or other stuff >

OK

    $ make manifest PACKAGES="-dnsmasq dnsmasq-full" | grep dnsmasq
    < should show just dnsmasq-full >

ERROR: unable to select packages:
  dnsmasq-full (no such package):
    required by: world[dnsmasq-full]
make[2]: *** [Makefile:227: package_install] Error 1
make[1]: *** [Makefile:167: _call_manifest] Error 2
make: *** [Makefile:345: manifest] Error 2

    $ make manifest PACKAGES="-dnsmasq" | grep dnsmasq
    < should be empty, no dnsmasq variants at all >

OK


Maybe something wrong in `make menuconfig`

efahl commented:
> Maybe something wrong in `make menuconfig`

Yes, oops on my part.  You need to go into menuconfig and find the variants/alternatives to ensure that they are built.

Are you familiar with searching in menuconfig?  Just type `/dnsmasq-full` and it should show you the package's location in the tree.   One of the entries usually has a number like `(1) -> dnsmasq-full...` and if you press `1` it will jump you right to it...
